orttraining/orttraining/training_api/checkpoint.cc

```diff
@@ -1,581 +1,765 @@
+
+      fbs::ParameterOptimizerStateBuilder optimizer_state_builder(builder);
+      optimizer_state_builder.add_param_name(builder.CreateString(param_name));
+      optimizer_state_builder.add_momentums(builder.CreateVector(momentums));
+
+      flatbuffers::Offset<fbs::ParameterOptimizerState> fbs_optimizer_state = optimizer_state_builder.Finish();
+      optimizer_states.emplace_back(std::move(fbs_optimizer_state));
     }
 
-    // Storing group-wise properties.
-    PropertyBag properties;
-    properties.AddProperty(builtin_lr_property_name, group_optimizer_state_ptr->initial_lr);
-    properties.AddProperty(builtin_step_property_name, group_optimizer_state_ptr->step);
-    std::vector<ONNX_NAMESPACE::TensorProto> group_wise_properties_tensor_protos;
-    properties.ToTensorProtos(group_wise_properties_tensor_protos);
+    const auto fbs_group_name = builder.CreateString(group_name);
+    const auto fbs_optimizer_states = builder.CreateVector(optimizer_states);
+
+    fbs::OptimizerGroupBuilder optimizer_state_builder(builder);
+    optimizer_state_builder.add_group_name(fbs_group_name);
+    optimizer_state_builder.add_initial_learning_rate(group_optimizer_state_ptr->initial_lr);
+    optimizer_state_builder.add_step(group_optimizer_state_ptr->step);
+    optimizer_state_builder.add_optimizer_states(fbs_optimizer_states);
+    const auto fbs_optimizer_group = optimizer_state_builder.Finish();
+    fbs_optimizer_groups.emplace_back(std::move(fbs_optimizer_group));
+  }
+
+  return Status::OK();
+}
+
+/**
+ * @brief Save from user defined properties to a flatbuffer checkpoint property bag.
+ *
+ * @param property_bag user defined properties.
+ * @param builder Flatbuffer builder.
+ * @param fbs_property_bag Flatbuffer property bag to be populated.
+ * @return Status of the operation.
+ */
+Status FromPropertyBag(const PropertyBag& property_bag, flatbuffers::FlatBufferBuilder& builder,
+                       flatbuffers::Offset<fbs::PropertyBag>& fbs_property_bag) {
+  if (property_bag.Size() == 0U) {
+    return Status::OK();
+  }
 
-    WriteTensorProtoToFile(
-        GetTensorProtoPropertiesFilePath(checkpoint_path, cur_group_filename_prefix),
-        group_wise_properties_tensor_protos, "[param_group_properties]");
+  InlinedVector<std::string> sorted_property_names;
+  sorted_property_names.reserve(property_bag.Size());
+  for ([[maybe_unused]] const auto& [name, value] : property_bag) {
+    sorted_property_names.push_back(name);
+  }
+  std::sort(sorted_property_names.begin(), sorted_property_names.end());
+
+  std::vector<flatbuffers::Offset<fbs::IntProperty>> ints;
+  std::vector<flatbuffers::Offset<fbs::FloatProperty>> floats;
+  std::vector<flatbuffers::Offset<fbs::StringProperty>> strings;
+  for (const auto& name : sorted_property_names) {
+    const auto& value = property_bag.GetProperty<PropertyDataType>(name);
+    const auto fbs_property_name = builder.CreateString(name);
+    if (std::holds_alternative<int64_t>(value)) {
```

Comment:
```suggestion
    if constexpr  (std::holds_alternative<int64_t>(value)) {
```

nit: can be covered later.

Reply:
I am not sure I follow how this can be `if constexpr`. From my understanding, the if condition is not compile-time evaluable.
I may have misunderstood your suggestion perhaps.

Reply:
i am wrong..